### DISCUSSION TITLE
Future of icalendar, looking for maintainer

### DISCUSSION THREAD

**Discussion by mauritsvanrees** [collective/icalendar#360] [Ideas]
I have created a [post on the Plone community forum](https://community.plone.org/t/future-of-icalendar-looking-for-maintainer/15445) about the future of this package, and needing a maintainer for it. If you want to add your thoughts, please create an account and comment there. But since `icalendar` seems to have users outside of the Plone community, where it originated, we can try out Discussions in this repository. I don't think I have used GitHub Discussions yet anywhere, but let's see.  Let me copy the entire post below.

The `icalendar` package in the Plone collective has been designated as a "critical" project on PyPI, meaning it is in the top 1 percent of downloaded packages. It was started by Martijn Faassen in 2005, saw lots of commits by Johannes Raggam and Rok Garbas around 2011/2012, during the time `plone.app.event` was created and added to Plone. You can see [all contributors](https://github.com/collective/icalendar/graphs/contributors).

Apparently I am the number four contributor, but that is only because I occasionally make a release. I hardly know anything about the inner workings of this package. As Plone Release Manager I *am* interested in having stable releases of this.

It would help if we get some fresh developer/maintainer blood in here. Is anyone capable and interested?

I have some ideas for the near future, to hopefully get things unstuck:

- We should no longer test this on the Plone Jenkins server. That would be something to arrange in [`mr.roboto`](https://github.com/plone/mr.roboto). I can have a look at that.
- The package can use Github Actions instead.
- According to `setup.py` the package supports Python 2.7 and 3.4-3.10. We should get GitHub actions to work with those versions. I expect that this is not really doable for at least Python 3.4 and 3.5, so we would need a new major release in which we drop support for some Python versions.
- Ideally there should be a branch that keeps working on all Python versions that are supported for Plone 5.2, which means: Python 2.7, 3.6, 3.7, 3.8. Maybe that just needs to be the current branch, which we then would copy to 4.x. If we fail to get Github Actions to work for this, Jenkins for Plone 5.2 should still keep testing the package.
- Increase master branch version to 5 (or 6 if needed). This should keep working on Plone 6.0 (which is near beta), which means Python 3.7-3.10.

So the **first action item** is really: try to get **Github Actions** (and locally `tox`) running on as much Python versions as possible, with a reasonable amount of effort.

Some links:

- PyPI: https://pypi.org/project/icalendar/
- Documentation: https://icalendar.readthedocs.io/en/latest/
- Code: https://github.com/collective/icalendar, where you can see it has 138 forks and 783 stars.
- Issue tracker: https://github.com/collective/icalendar/issues
- 18 open pull requests: https://github.com/collective/icalendar/pulls, including:
  - My failed attempt at [adding GitHub actions](https://github.com/collective/icalendar/pull/340), from November 2021.
  - [Document Setup process for development](https://github.com/collective/icalendar/pull/358).

**Comment by mauritsvanrees** [comment]
I have managed to revive my gh-actions PR, and it is now green and merged. I have disabled Jenkins tests in the settings in this repo. Something more might be needed, but for now I think on new PRs these checks should no longer pop up.

I have released 4.1.0 and 5.0.0a1. The only current difference is that 4.1.0 (new branch 4.x) only drops some Python versions from the tests, and on 5.0.0a1 (master) they are really officially not supported anymore. We could drop more versions there if we want. Now at least both branches have a reasonable tox and gh-actions.  So for clarity:

- Branch 4.x is for Python 2.7, 3.6-3.10, and PyPy3. Those are tested. Technically it should still work on 3.4, 3.5 and PyPy2 and let's try not to break this.
- The master branch is tested and officially supported on Python 2.7, 3.6-3.10, and PyPy3. I would be in favour of dropping 2.7 and 3.6 here.

**Reply by geier** [reply]
All for dropping old version of python

**Comment by niccokunzmann** [comment]
@mauritsvanrees I am maintaining [recurring-ical-events](https://github.com/niccokunzmann/python-recurring-ical-events) and others that build on top of `icalendar` in my free time. I will give it a go for a bit but I am unsure about the amount of time that I can put in and how the trust hand-over works. This is a really critical project. That should be made clear also in the distribution of the new versions.

Is it alright to raise a `Warning` in the latest version that a maintainer is needed? I mean - this should concern people using it really! I am a bit more concerned as this did not receive any comment until now..

After adding #365, I see that recurring-ical-events accounts for about a bit less than 1/3 of the installation reasons for this package.

While I am not in the depth of the code of this, I have long-term Python experience and read Python-libraries for fun. Also, I can keep the discussion going forward while keeping the quality of the code. Like this is a no-brainer: https://github.com/collective/icalendar/pull/330 and it just needs some attention - for a year now...
If I get write access, I will ping you for checking my PRs (mostly docs/tests) while keeping some of the other ones going forward...
What are your thoughts on that?

**Reply by geier** [reply]
Please don't emit a Warning at least not during run time (while installing, e.g. in setup.py is fine for me), end user applications only will need to suppress that. 

Recurring-ical-events looks interesting, I'll give it a go and see if it can replace a lot of ugly code for us.

**Reply by niccokunzmann** [reply]
> Recurring-ical-events looks interesting, I'll give it a go and see if it can replace a lot of ugly code for us.

Maybe you will find some edge cases.

**Comment by geier** [comment]
Thank you for inviting me @mauritsvanrees to become a maintainer. I'd love to contribute more to icalendar, but I rarely have the time, but I'm happy to review some PRs and contribute occasionally (but mostly if it concerns [khal](http://github.com/pimutils/khal)).

Also, I think somebody more understanding of the needs of the Plone community would probably make sense.

I'd like to not only drop python 3.6 but also 3.7 (because python 3.8 makes type hinting really a lot easier IMHO), make icalendar always return `str`s not `byte`s, make sure we always return valid icalendar objects, migrate from `pytz` to `tzinfo`. All of which would break with the long-term stability that -- if I understand it correctly -- plone needs.

**Comment by mauritsvanrees** [comment]
Thank you @niccokunzmann and @geier for stepping up, and be willing to do what you can. Let's see how it goes.

For me it is okay to drop Python 2.7 and 3.7 from the master branch. I have changed Plone 5.2 to use branch 4.x. And on latest Plone 6.0 I have added a pin for Python 3.7 to use the latest 4.x version as well. I have created https://github.com/collective/icalendar/pull/366 to drop older versions. Can you review.

I have merged several PRs today.
Ping me if you have questions related to Plone, or need to get things unstuck.

Christian already has PyPI rights. Nicco not yet. I see two accounts on https://pypi.org/project/recurring-ical-events/. Which account can I add for you?

**Reply by niccokunzmann** [reply]
I do not much mind not having access to PyPI at the moment. I could probably contribute more if I had write access to this repository.

**Reply by mauritsvanrees** [reply]
Ah, I did not realise you do not have write access yet. Everyone can get write access to the [Plone Collective](https://collective.github.io) where this package lives. I can give you an account there. Or if you you prefer I can add you to this repository only.

**Comment by niccokunzmann** [comment]
> Nicco not yet. I see two accounts on https://pypi.org/project/recurring-ical-events/. Which account can I add for you?

@mauritsvanrees, The account to add would be `niccokunzmann` - the other one is an old account for automatic deployments before API tokens.

**Reply by mauritsvanrees** [reply]
I have sent you an invite.

**Reply by niccokunzmann** [reply]
Thanks. I see that some people do not have 2 factor authentication enabled. Would it be alright to contact them and ask if they like to do that or be removed from the project? There are many people who do have it enabled. That might be low priority, though, for now.

**Reply by mauritsvanrees** [reply]
One of these users is `plone`, which is a shared account of the Plone release managers (including me). Enabling 2FA there sounds tricky.

Another is `maxm`, who I do not know.

I do know the two remaining others, who I will ping here.

**Reply by mauritsvanrees** [reply]
@garbas @rnixx You are among the owners of `icalendar`. It turns out that this package is so popular that it has been tagged as critical by PyPI. This means PyPI wants all contributors to use two-factor authentication. Some [info here](https://pypi.org/security-key-giveaway/). You do not have 2FA enabled on PyPI. Would you like to do that? If you are no longer interested in PyPI ownership I could also remove you there.

**Reply by garbas** [reply]
Hi @mauritsvanrees! Feel free to remove me from the owners list. I'm not doing much of python development these days. Happy coding to all!

**Reply by mauritsvanrees** [reply]
Hi Rok. I have removed you. Happy coding to you too. And you know: there is always a place in the Python and Plone communities for you!

**Comment by niccokunzmann** [comment]
@geier, I went through all the pull requests and updated people there about the status of the repo and asked them to rebase/merge of possible.

**Comment by WhyNotHugo** [comment]
Happy to see this under active maintenance again! I'm using this library for `todoman`, and currently support Python 3.10. I'd like to switch to moving `ZoneInfo` in future, which is part of Python's stdlib but has a backport for 3.8 and 3.9. Any thoughts on this?

In the meantime, I've rebased my old PR which removes a lot of leftovers from older python versions: https://github.com/collective/icalendar/pull/327

And also this other one just auto-generated with `pyupgrade --py38-plus`: https://github.com/collective/icalendar/pull/376

**Reply by niccokunzmann** [reply]
Thanks for that!

(1) I see your PRs and I think I would want to wait until 3.7 or not 3.7 is decided to merge 3.8 upgrades - I commented there as you can see.
(2) **Personally**: I test pytz and zoneinfo for python-recurring-ical-events and pytest allows to run all tests on both implementations after another. BUT: moving forward is dropping pytz probably as it is much better to use zoneinfo and people use pytz in the wrong way so many times. Python3.7 EOL is in 10 months - good question. What do you think? I think, we discuss it in #368 at the moment.

**Reply by stevepiercy** [reply]
@niccokunzmann, @thet and I have started the planning phase of work to update the Plone Event content type, which primarily uses `plone.app.event` which in turn has `icalendar` as one of its dependencies. We also decided to transition from `pytz` to `zoneinfo`, and our work is targeted for Plone 6.2 or 7.0, whichever version drops support for Python 3.8 first. You can read our Plone Improvement Proposal (PLIP) at https://github.com/plone/volto/issues/5629.

I lead the Plone Documentation Team, and developed a timezone converter application about 15 years ago in an obscure web programming language before I started using Python. I started watching this repo. I hope to contribute to this project, especially documentation. I also helped @mauritsvanrees revise the [Plone release policy](https://plone.org/download/release-schedule) for 2024.

**Reply by niccokunzmann** [reply]
@stevepiercy,
I acquired funding from NLNet #595 and zoneinfo for icalendar is on the list #609. Cool that you are also going this way!

I would discuss how to go about moving to zoneinfo in #609 instead of here. Please leave a comment on what you see needs doing. Thanks for linking https://github.com/plone/volto/issues/5629.

Over a year ago, we started refactoring the tests so that we could potentially allow to use pytz, zoneinfo and other timezone implementations in icalendar - make it flexible. Moving to zoneinfo, nobody has objected to drop pytz. So, being flexible might not be needed. Thanks for your note here in this project and I look forward to seeing how we can join forces!

**Reply by niccokunzmann** [reply]
Let's continue this conversation here: #609.

**Comment by niccokunzmann** [comment]
For those who do not want to read the full discussion, a summary (I will edit it here), please leave a comment.

# SUMMARY

A list of the maintainers of the package with their abilities/focuses 

- @geier - likes to contribute code and evolve the package?  
    Abilities:
    - [x] review
    - [x] merge
    - [x] deploy new versions
    - [x] GitHub admin
- @niccokunzmann - likes to involve people, document, test.  
    Abilities:
    - [x] review
    - [x] merge
    - [x] deploy new versions
    - [x] GitHub admin
- @mauritsvanrees - likes to make sure this project is maintained and progressing.
    Abilities:
    - [ ] review
    - [x] merge
    - [x] deploy new versions
    - [x] GitHub admin

Does this sound true? Did I forget someone?

**Comment by niccokunzmann** [comment]
@geier - there are branches that are very old...
![image](https://user-images.githubusercontent.com/564768/185740370-aec4b216-0654-4640-8103-3ada7f64e0d7.png)
The latest ones (5 years) are your branches - I would hope the others can safely be deleted. What do you think since you are more involved in these branches?

**Reply by geier** [reply]
yes, we can delete those.

**Comment by WhyNotHugo** [comment]
Is buildout still actively used for this project? If not, `buildout.cfg` and `bootstrap.py` can be deleted.

**Reply by niccokunzmann** [reply]
Hm. Email may not work on discussions...
So, I do not know what these are, never used these. I think, you can remove them - if we need them, we can revert the PR. Also, it removes clutter that does not add to functionality. There is nothing in the documentation, only in setup.cfg.

**Reply by mauritsvanrees** [reply]
`buildout.cfg` and `bootstrap.py` can be removed indeed.

**Comment by niccokunzmann** [comment]
I allowed the actions to run as they do not contain any secrets and it is nice to get the result and not wait for a maintainer...
![image](https://user-images.githubusercontent.com/564768/185805398-ebccb04a-ca20-4a24-8c50-626637616a7a.png)

**Comment by niccokunzmann** [comment]
@geier @mauritsvanrees, I have this proposal:

The intention of this is to smooth out the way to become a maintainer & contributor in this project.

The strategy:
- We create two teams:
    - maintainers (admin access)
    - contributors (write access, should not merge their own PR)
- When people create a successful pull request, we ask them if they would like write access to the repository. If they agree, we add them to the contributors group.

Hoped Result:
- people are more engaged as they can review and merge other PRs
- we clarify the process and documentation when we see something happening that we do not like
- when people see that they can review/merge - after a while they might like helping out more
- we get workload off our shoulders, more reviews, more discussion, more engaged community.

What do you think, should we try it out?

**Reply by mauritsvanrees** [reply]
Sounds fine to me.

**Reply by niccokunzmann** [reply]
@mauritsvanrees I am not allowed to create the icalendar-admin and the icalendar-contributor teams in this organization.
![image](https://user-images.githubusercontent.com/564768/188206036-cbe646cb-cf97-4835-bd19-a75ac4c7d6db.png)

I would like to add @jacadzaca to the contributor team because of the interest in maintenance as shown in #392.
Can you or who could

- create the two teams
- make @geier and @niccokunzmann admin of the two teams?

thanks for your support!

**Reply by mauritsvanrees** [reply]
I have added the two teams. Sorry it took a while.
I wondered why you could not do this, since you already have admin rights. But they had to be added on the organisation level.
`icalendar-admin` has admin rights, `icalendar-contributor` has write rights. I have invited you and @geier to the admin group.

**Comment by niccokunzmann** [comment]
Thank you very much, no problem!

**Comment by niccokunzmann** [comment]
I invited @angatha to be come a contributor as I saw the contributions in the last days reviewing PRs.
Invitation: https://github.com/collective/icalendar/pull/413#issuecomment-1265768631

**Reply by SashankBhamidi** [reply]
Hey @angatha!

I wanted to invite you to join our Matrix group for the iCalendar community. I think it'd be great to have you there for discussions, collaboration, and project updates!

Join here: https://matrix.to/#/#pycal:chat.pycal.org

Looking forward to seeing you there!

**Comment by niccokunzmann** [comment]
I would like to give @jacadzaca equal maintainer status to myself. The work is nice together, committed and thoughtful. @jacadzaca published version 5.0.1. Thanks! @jacadzaca do you accept? Does anyone have any comments?

Here is a blog post that inspires me: http://hintjens.com/blog:95

**Reply by mauritsvanrees** [reply]
Sounds good to me! I am happy to see a community evolving here, with more maintainers.

**Reply by jacadzaca** [reply]
@niccokunzmann I'm OK with being a maintainer, what new privileges will I get?

**Reply by niccokunzmann** [reply]
- [x] Repository Admin
- [x] Maintainer on PyPI
- [x] Maintainer on readthedocs
- [x] PyPI env access

see https://icalendar.readthedocs.io/en/latest/maintenance.html#maintainers

@jacadzaca 
What is your PyPI username? Do you have 2 factor authentication enabled?
What is your readthedocs username?

**Reply by jacadzaca** [reply]
@niccokunzmann My pypi username is [jacadzaca](https://pypi.org/user/jacadzaca/) and I have 2FA enabled.

**Reply by niccokunzmann** [reply]
PyPI invite is pending. Could you create an account on readthedocs? I would like one more person there to improve the bus factor.

**Reply by jacadzaca** [reply]
I created an account on readthedocs, my username is jacadzaca.

**Reply by niccokunzmann** [reply]
Thanks, I sent an invitation there. You are 100% set up with authority and responsibility! :partying_face:

**Comment by niccokunzmann** [comment]
That is documented above the release process. I will do it one-by-one when I take the time.

**Comment by niccokunzmann** [comment]
I do not really know where to post this but here is the usage statistics from readthedocs. I think, this is nice to keep in mind when it comes to making the documentation nicer.
![image](https://user-images.githubusercontent.com/564768/201549461-1449ca60-2a36-4e59-8e05-564e79b5f9ac.png)

**Comment by niccokunzmann** [comment]
I am adding @ennamarie19 as a contributor with write access so we can be taught about the fuzzing tests and also have more people to look after the project who are more committed. 
See also: https://github.com/collective/icalendar/pull/574#issuecomment-1789367456

**Reply by niccokunzmann** [reply]
@ennamarie19, welcome!

**Comment by ennamarie19** [comment]
Thanks so much! Looking forward to what comes of this!

**Comment by niccokunzmann** [comment]
@SashankBhamidi you worked as a contributor for a while now. Would you like to become a maintainer?
https://icalendar.readthedocs.io/en/latest/contribute/maintenance.html#icalendar-maintainers

**Reply by SashankBhamidi** [reply]
@niccokunzmann Thanks for the offer. I really appreciate the trust.

I haven’t maintained a project before, so I’d love to better understand what the expectations and responsibilities would look like. I’ve been learning a lot through contributing here and from you and Steve, and I’d definitely like to continue growing.

Could you share what being a maintainer typically involves in practice?

**Reply by niccokunzmann** [reply]
It is mostly what you do as a contributor when it comes to writing code.
The focus changes though: You are more responsible for the project living on than a change being introduced - even if it is your change.
I think: Our communication is good which allows you to grow into what is needed based on feedback. Also, I am away in May and you might want a release then. Being a maintainer now, you don't have to do anything but it would be good to issue a release once before May.

You can then
- create new releases
- administrate icalendar

What do you think?

**Reply by SashankBhamidi** [reply]
Thanks for clarifying. I’m open to becoming a maintainer and handling a release before May. My availability may fluctuate since I’m building my company, and if that ever becomes an issue, I’d step back and communicate it. Let me know if that works.

**Reply by niccokunzmann** [reply]
Yes, that works. It is not intended as a burden but by having more people who can do the work, we will have a better response-ability as a team and project.

**Reply by stevepiercy** [reply]
In addition, there are 8 total maintainers on PyPI now. If ever you feel overwhelmed or become unavailable, one of them could step in.

Plone, especially, depends on icalendar. Either @mauritsvanrees, @thet, and @tisto from Plone could step in. Or dragged in. 😉 @mauritsvanrees is the Plone release manager.

I assume that the users on PyPI @collective and @plone are owned by the Plone Admin Team as well, and could make a release. @mauritsvanrees is on that team, and might enlighten us about those two users and their powers. I have no insight there.

I am an owner on both the Plone and collective GitHub organizations, which grants me Admin access.

I don't have PyPI maintainer access for icalendar. I'd like not to have the honor. When I took on PyPI maintainership of sphinxcontrib-httpdomain, I didn't realize that it had 1/3 the downloads of icalendar, which is still a *lot*, but no one else stepped up, and it needed to be done.

**Reply by SashankBhamidi** [reply]
Thanks both. I really appreciate the support and clarity around this.

**Reply by niccokunzmann** [reply]
@SashankBhamidi Welcome to the maintainer's team. Please view this: https://github.com/collective/icalendar/pull/1195

**Reply by mauritsvanrees** [reply]
* PyPI plone org: me, Érico Andrei, Eric Steele, Fred van Dijk, Gil Forcada.
* PyPI collective org: those same people, plus Godefroid Chapelle, Jens Klein, and whoever has access to the `collective` user (this was a shared account, used when PyPI orgs did not yet exist, at least the Plone admins team will have access).

So basically a bunch of people who have been active a lot in Plone.

**Comment by niccokunzmann** [comment]
I welcome @lcampanella98 as a contributor to icalendar!

Thanks for being ready to pitch in on a voluntary base! No pressure, just where it fits.

See the discussion here:
https://github.com/collective/icalendar/pull/1400#issuecomment-4729414513

You have write access, so that you can merge pull requests of other contributors if you feel comfortable to do so.
Feel free to put in a PR for the CODEOWNERS file if you want to be notified only for PRs to specific files.

Happy contributing! Let us know if you have any questions :heart: I look forward to working along side you and growing a community of collaborators to make icalendar the best Python library for calendaring! :rocket:

PS: You need to accept the invite.

**Reply by lcampanella98** [reply]
Thanks for the warm welcome! Happy to pitch in where it fits. I've accepted the invite, and I'll put in a CODEOWNERS PR if I decide to scope notifications later. Looking forward to working with you all!
